### DISCUSSION TITLE
why irrep is negative, why direct product of irreps not associative?

### DISCUSSION THREAD

**Discussion by yuhanzheng1** [pyscf/pyscf#2680] [Q&A] [answered]
The symmetry of one molecule is 'Coov'
The irreps ID of mo-orbitals are [0, 10, 11] respectively. #where irreps = symm.label_orb_symm(mol, mol.irrep_id, mol.symm_orb, mf.mo_coeff)

Why irreps[1] × irreps[2] × irreps[2]=-12?   Why （irreps[1] × irreps[2] )×irreps[2] =-12     not equal to     irreps[1]× (irreps[2] × irreps[2])=  irreps[1] ×Ag= irreps[1]=10?

**Comment by yuhanzheng1** [comment]
'the code is as follows'
import pyscf
from pyscf import gto
import numpy as np
import scipy
import openfermion as of
from pyscf import symm

basis = {'C':  [[0, 
    [4231.61, 0.002029], 
    [634.882, 0.015535], 
    [146.097, 0.075411],
    [42.4974,0.257121],
    [14.1892,0.596555],
    [1.9666,0.242517]],
    [0,[5.1477,1]],
    [0,[0.4962,1]],
    [0,[0.1533,1]],
    [0,[0.015,1]],
    [1,[18.1557,0.018534],[3.9864,0.115442],[1.1429,0.386206],[0.3594,0.640089]],
    [1,[0.1146,1]],
    [1,[0.011,1]],
    [2,[0.75,1]]],
    'H': [[0,
    [19.2406,0.032828],
    [2.8992,0.231208],
    [0.6534,0.817238]],
    [0,[0.1776,1]],
    [0,[0.025,1]],
    [1,[1,1]]]
    }

geom = [['C',(0, 0, 0)],['H',(0, 0, 2.26184115)]] 
charge = 1
n_process=16
groupname = 'Coov'
freezon = [[0,1,2,3], [6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27, 28, 29, 30, 31, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49]]
mol = pyscf.gto.M(atom = geom, basis = basis, spin = 0, charge = charge,symmetry=True)

'irreps~~~'
mol.build() 
print('mol_build is ok')

mf = pyscf.scf.RHF(mol)
hf_energy = mf.kernel()
print('hf_energy is ok')

print('mo_energy',mf.mo_energy)
n_orb = (mf.mo_coeff).shape[1]  # molecule.nao_nr()
n_orb_occ = sum(mol.nelec) // 2
irreps = symm.label_orb_symm(mol, mol.irrep_id, mol.symm_orb, mf.mo_coeff)
if freezon[0] is not None and \
		freezon[1] is not None:
	n_orb_ori = len(irreps)
	except_indices_spin = freezon[0] + \
		freezon[1]
	irreps = [irreps[i] for i in range(n_orb_ori) if
			  (2 * i not in except_indices_spin and
			   2 * i + 1 not in except_indices_spin)]
print('after_freezon_orbsym',irreps)



' calculate irreps[1] × irreps[2] × irreps[2]'
irrep_rep = np.array([0])
print('irrep_rep0',irrep_rep)
print('irreps',irreps)
for ii in [1,2,2]: 
    print('ii,irreps[ii]',ii,irreps[ii])                                                                  
    irrep_rep = pyscf.symm.direct_prod(irrep_rep,np.array([irreps[ii]]),groupname=groupname)[0]
    print('irrep_rep',irrep_rep)  
irrep_rep = irrep_rep[0]
print('irrep_rep',irrep_rep)

**Comment by yuhanzheng1** [comment]
Hello, can someone help me? Can we use pyscf to calculates the direct product of irreps under ‘Coov’ symmetry?

**Comment by sunqm** [comment, marked as answer]
The direct product in your case involves is the product of 2D irreps. It can be decomposed to 2 1D irreps and 1 2D irreps. The `direct_prod` function only outputs one of the three irreps. You cannot use this function to find the accurate direct products for cylindrical symmetry.

To accurately handle the symmetry of cylindrical symmetry, the orbitals should be linearly transformed to (+) and (-) complex orbitals from the real orbitals in x, y representations. Then you can obtain the momentum of the orbitals and use the momentum to derive the direct product. In fci module, the cylindrical symmetry is treated this way, e.g.
https://github.com/pyscf/pyscf/blob/231da24f746d3a7988a4c48c9237c4788bda6a36/pyscf/lib/mcscf/fci_contract.c#L983

**Comment by verena-neufeld** [comment]
Hi @yuhanzheng1 , thanks for asking. Has your question been answered?
